File: Sorting/Counting Sort.cs
/// <summary>
/// Counting sort
/// 
/// Time complexity: O(n + k)
/// Space complexity: O(n + k)
/// 
/// see this link for illustration: https://www.programiz.com/dsa/counting-sort
/// </summary>

namespace SearchingAndSorting
{
    internal class Counting_Sort
    {

        static void CountSort(int[] arr)
        {
            int n = arr.Length;
            int[] output = new int[n];

            int max = arr.Max();

            // store the count of the each array elements in a count array
            int[] count = new int[max + 1];

            // initialize all the array elements in count array as 0
            Array.Fill(count, 0);

            // store the frequency/count of each elements of given array in count array
            for (int i = 0; i < n; i++)
                count[arr[i]]++;

            // store the sum of current element and previous element in each index
            for (int i = 1; i <= max; i++)
                count[i] += count[i - 1];

            // iterate the original array in reverse order
            // we ll take the value from original array and compare with the count array index
            // at that particular index in count array, we take the value as the index for the output array
            // and index value as the actual value, and we decrement value in count array
            for (int i = n - 1; i >= 0; i--)
            {
                output[count[arr[i]] - 1] = arr[i]; // either do a right shift by 1 in count array or subtract by 1 here
                --count[arr[i]];
            }

            printSortedArray(output);
        }

        static void printSortedArray(int[] arr)
        {
            for (int i = 0; i < arr.Length; i++)
            {
                Console.Write(arr[i] + " ");
            }
        }

        public static void Main(string[] args)
        {
            int[] arr = new int[] { 1, 3, 4, 4, 3, 1, 1, 2, 2, 1 };

            CountSort(arr);
        }
    }
}

File: Sorting/counting-sort.ts
function CountingSort(arr: number[]) {

    const max = Math.max(...arr);
    let count = new Array(max + 1), ans = new Array(arr.length);
    count.fill(0);

    // create the count array
    for (let i = 0; i < arr.length; i++) {
        count[arr[i]]++;
    }


    // compute the prefix sum in order to determine the number of elements and its correct position
    for (let i = 1; i < count.length; i++) {
        count[i] = count[i] + count[i - 1];
    }

    // place the elements in correct position using prefix sum
    for ( let i = arr.length - 1; i >= 0; i-- ) {
        ans[--count[arr[i]]] = arr[i];
    }

    console.log(ans);
}

const arr1 = [4, 2, 2, 3, 1, 6, 4, 6, 9, 9, 6];
CountingSort(arr1);


// Time Complexity: O(N+M) in all cases, where N and M are the size of inputArray[] and countArray[] respectively.
// Auxiliary Space: O(N+M), where N and M are the space taken by outputArray[] and countArray[] respectively.
